### PR TITLE
On demand

### PR DESCRIPTION
**Pull request by PrasannaLanka**

### CODE REVIEW

File: src/libIR2Vec.cpp [hunk 1]
```diff
@@ -9,30 +9,31 @@
...
   if (mode == IR2Vec::IR2VecMode::FlowAware) {
     IR2Vec_FA FA(M);
     FA.generateFlowAwareEncodings(o);
     instVecMap = FA.getInstVecMap();
     funcVecMap = FA.getFuncVecMap();
     pgmVector = FA.getProgramVector();
-  }
-
-  else if (mode == IR2Vec::IR2VecMode::Symbolic) {
+  } else if (mode == IR2Vec::IR2VecMode::Symbolic && !fname.empty()) {
```
Comment (ShikharJain97) [inline src/libIR2Vec.cpp]: suppose user doesnt gave any function name and want symbolic embeddings, will this work?

**Review by ShikharJain97 — COMMENTED** [review]
code seems to be correct for on demand symbolic encoding generation.

File: src/FlowAware.cpp [hunk 2]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+  auto tmpParent = func2Vec(*function, funcStack);
+  // funcVecMap is updated with vectors returned by func2Vec
+  funcVecMap[function] = tmpParent;
+  auto calledFunctions = funcCallMap[function];
+  for (auto &calledFunction : calledFunctions) {
+    if (!calledFunction->isDeclaration() &&
+        visitedFunctions.count(calledFunction) == 0) {
+      // doing casting since calledFunctionsis of type of const llvm::Function*
+      // and we need llvm::Function* as argument
+      Function *foo = const_cast<Function *>(calledFunction);
```
Comment (svkeerthy) [inline src/FlowAware.cpp:230]: Use auto, give better variable name and change const_cast.

File: src/FlowAware.cpp [hunk 3]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+// newly added
+// This function will update funcVecMap by doing DFS starting from parent
+// function
+void IR2Vec_FA::updateFuncVecMap(
+    llvm::Function *function,
+    llvm::SmallMapVector<const llvm::Function *,
+                         llvm::SmallVector<const llvm::Function *, 10>, 16>
+        &funcCallMap,
+    std::unordered_set<const llvm::Function *> &visitedFunctions) {
+  visitedFunctions.insert(function);
```
Comment (svkeerthy) [inline src/FlowAware.cpp:218]: Please add an assertion on function for nullptr

File: src/FlowAware.cpp [hunk 4]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+
+// newly added
+// This function will update funcVecMap by doing DFS starting from parent
+// function
+void IR2Vec_FA::updateFuncVecMap(
+    llvm::Function *function,
+    llvm::SmallMapVector<const llvm::Function *,
+                         llvm::SmallVector<const llvm::Function *, 10>, 16>
+        &funcCallMap,
+    std::unordered_set<const llvm::Function *> &visitedFunctions) {
```
[the remark this answers is not in the document]
Comment (svkeerthy) [reply src/FlowAware.cpp:217]: Can we make the argument as const?

File: src/FlowAware.cpp [hunk 5]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+      updateFuncVecMap(foo, funcCallMap, visitedFunctions);
+    }
+  }
+}
+
+void IR2Vec_FA::generateFlowAwareEncodingsForFunction(
+    std::ostream *o, std::string name, std::ostream *missCount,
+    std::ostream *cyclicCount) {
+  // clock_t start = clock();
+  collectWriteDefsMap(M);
```
Comment (svkeerthy) [inline src/FlowAware.cpp:242]: Can this also be moved to constructor and removed from `generateFlowAwareEncodings()` as well?

File: src/FlowAware.cpp [hunk 6]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+    llvm::ItaniumPartialDemangler Mangler;
+    if (Mangler.partialDemangle(mangled)) {
+      Result = &demangledName[0];
+    } else {
+      Result = Mangler.getFunctionBaseName(Buf, &Size);
+    }
+    if (!f.isDeclaration() && Result == name) {
+      // If fname is matched with one of the functions in module, we will update
+      // funcVecMap of it and it's child functions recursively
+      std::unordered_set<const Function *> visitedFunctions;
```
Comment (svkeerthy) [inline src/FlowAware.cpp:267]: Please change this to appropriate LLVM set container

File: src/FlowAware.cpp [hunk 7]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+    std::ostream *cyclicCount) {
+  // clock_t start = clock();
+  collectWriteDefsMap(M);
+  int noOfFunc = 0;
+  for (auto &f : M) {
+    auto funcName = f.getName().str();
+    // getting demangled function name
+    std::size_t sz = 17;
+    int status;
+    char *const readable_name =
```
Comment (svkeerthy) [inline src/FlowAware.cpp:249]: Please create a function to get demangled method name and call it in both the places.

Comment (svkeerthy) [reply src/FlowAware.cpp:249]: Also appropriately change the existing logic in `generateFlowAwareEncodings()` to make the call to the same function.

File: src/FlowAware.cpp [hunk 8]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+    if (!f.isDeclaration() && Result == name) {
+      // If fname is matched with one of the functions in module, we will update
+      // funcVecMap of it and it's child functions recursively
+      std::unordered_set<const Function *> visitedFunctions;
+      updateFuncVecMap(&f, funcCallMap, visitedFunctions);
+    }
+  }
+  // iterating over all functions in module instead of funcVecMap to preserve
+  // order
+  for (auto &f : M) {
```
Comment (svkeerthy) [inline src/FlowAware.cpp:273]: This logic is very similar to what we have in `generateFlowAwareEncodings()`. Please try creating a function and use it in both the places.

File: src/FlowAware.cpp [hunk 9]
```diff
@@ -211,6 +201,149 @@ void IR2Vec_FA::generateFlowAwareEncodings(std::ostream *o,
...
+      // errs() << "\n";
+      scaleVector(calleeVector, WA);
+      auto tmpParent = funcVecMap[&f];
+      std::transform(calleeVector.begin(), calleeVector.end(),
+                     tmpParent.begin(), tmpParent.begin(), std::plus<double>());
+      funcVecMap[&f] = tmpParent;
+    }
+  }
+
+  for (auto &f : M) {
```
Comment (svkeerthy) [inline src/FlowAware.cpp:292]: This logic is very similar to what we have in `generateFlowAwareEncodings()`. Please try creating a function and use it in both the places.

File: src/IR2Vec.cpp [hunk 10]
```diff
@@ -39,6 +39,9 @@ cl::opt<std::string> cl_iname(cl::Positional, cl::desc("Input file path"),
 
 cl::opt<std::string> cl_oname("o", cl::Required, cl::desc("Output file path"),
                               cl::cat(category));
+// for on demand generation of embeddings taking function name
+cl::opt<std::string> cl_fname("fname", cl::Optional, cl::init(""),
```
Comment (svkeerthy) [inline src/IR2Vec.cpp:43]: fname could be perceived as file name as well. Please consider renaming it appropriately.

File: src/IR2Vec.cpp [hunk 11]
```diff
@@ -113,8 +119,25 @@ int main(int argc, char **argv) {
     exit(1);
 
   auto M = getLLVMIR();
+  // newly added
```
Comment (svkeerthy) [inline src/IR2Vec.cpp:122]: Seems like the file descriptors are not closed even in the existing code. Can you please close them appropriately?

**Review by svkeerthy — CHANGES_REQUESTED** [review]

**Review by svkeerthy — APPROVED** [review]

**Pull request merged**